services/sharedchannel/sync_send.go

```diff
@@ -0,0 +1,262 @@
+		ChannelId: channelId,
+		Time:      scr.LastSyncAt,
+	}
+	posts, err := scs.server.GetStore().Post().GetPostsSince(opts, true)
+	if err != nil {
+		return err
+	}
+
+	var repeat bool
+
+	pSlice := posts.ToSlice()
+	max := len(pSlice)
+	if max > MaxPostsPerSync {
+		max = MaxPostsPerSync
+		repeat = true
+	}
+
+	if !rc.IsOnline() {
+		scs.notifyRemoteOffline(pSlice, rc)
+		return nil
+	}
+
+	msg, err := scs.postsToMsg(pSlice[:max], cache)
+	if err != nil {
+		return err
+	}
+
+	rcs := scs.server.GetRemoteClusterService()
+	if rcs == nil {
+		return fmt.Errorf("cannot update remote cluster for channel id %s; Remote Cluster Service not enabled", channelId)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), remotecluster.SendTimeout)
+	defer cancel()
+
+	err = rcs.SendMsg(ctx, msg, rc, func(msg model.RemoteClusterMsg, rc *model.RemoteCluster, resp remotecluster.Response, err error) {
+		if err != nil {
+			return
+		}
+
+		// TODO: Any Post(s) that failed to save on remote side are included in an array of Post ids in syncResponse[ResponsePostErrors].
+		//       Write ephemeral message to post author notifying for each post that failed.
+
+		// update SharedChannelRemote's LastSyncAt if send was successful
+		lastSync, err := resp.Int64(ResponseLastUpdateAt)
+		if err != nil || lastSync == 0 {
+			scs.server.GetLogger().Error("invalid last sync response after update shared channel",
+				mlog.String("remote", rc.DisplayName), mlog.Err(err), mlog.Any("last_update_at", resp[ResponseLastUpdateAt]))
+			return
+		}
+
+		if err := scs.server.GetStore().SharedChannel().UpdateRemoteLastSyncAt(scr.Id, lastSync); err != nil {
+			scs.server.GetLogger().Error("error updating LastSyncAt for shared channel remote", mlog.String("remote", rc.DisplayName), mlog.Err(err))
+		}
+	})
+
+	if err == nil && repeat {
+		contTask := newSyncTask(channelId, rc.RemoteId)
+		scs.addTask(contTask)
+	}
+	return err
+}
+
+// notifyRemoteOffline creates an ephemeral post to the author for any posts created recently.
+func (scs *Service) notifyRemoteOffline(posts []*model.Post, rc *model.RemoteCluster) {
+	// only send one ephemeral post per author.
+	notified := make(map[string]struct{})
+
+	// range the slice in reverse so the newest posts are visited first; this ensures an ephemeral
+	// get added where it is mostly likely to be seen.
+	for i := len(posts) - 1; i >= 0; i-- {
+		post := posts[i]
+		if _, ok := notified[post.UserId]; ok {
+			continue
+		}
+
+		postCreateAt := model.GetTimeForMillis(post.CreateAt)
+
+		if post.DeleteAt == 0 && post.UserId != "" && time.Since(postCreateAt) < NotifyRemoteOfflineThreshold {
+			T := scs.getUserTranslations(post.UserId)
+			ephemeral := &model.Post{
+				ChannelId: post.ChannelId,
+				Message:   T("sharedchannel.cannot_deliver_post", map[string]interface{}{"Remote": rc.DisplayName}),
+				CreateAt:  post.CreateAt + 1,
+			}
+			scs.app.SendEphemeralPost(post.UserId, ephemeral)
+
+			notified[post.UserId] = struct{}{}
```

Comment:
This is probably one of those cases where I would have used a `bool` just because it makes it almost obvious what the map is used for.

Reply:
Fixed via https://github.com/mattermost/mattermost-server/pull/16638/commits/cfe00799ae130542f7eaadc54df79a8d9f5e6bff